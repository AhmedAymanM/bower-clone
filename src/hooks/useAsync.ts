// README: ideally I would use react-query or similar library to manage server
// state and caching, I don't expect this hook to work perfectly but its is
// handy as a POC

import { useState, useEffect } from 'react'

export enum STATUS {
  IDLE = 'idle',
  LOADING = 'loading',
  SUCCESS = 'success',
  ERROR = 'error',
}

export function useAsync<T = unknown>(asyncFn: () => Promise<T>) {
  const [status, setStatus] = useState<STATUS>(STATUS.IDLE)
  const [data, setData] = useState<Awaited<T>>()
  const [error, setError] = useState<unknown>()

  useEffect(() => {
    setStatus(STATUS.IDLE)
    setData(undefined)
    setError(undefined)
    const fetchAsync = async () => {
      setStatus(STATUS.LOADING)

      try {
        const response = await asyncFn()
        setData(response)
        setStatus(STATUS.SUCCESS)
      } catch (error) {
        setError(error)
        setStatus(STATUS.ERROR)
      }
    }

    fetchAsync()
  }, [asyncFn])
  return { data, status, error }
}
