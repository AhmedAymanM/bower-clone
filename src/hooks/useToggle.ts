import { useCallback, useState } from 'react'

export function useToggle(defaultValue?: boolean): [boolean, () => void] {
  const [value, setValue] = useState(!!defaultValue)

  const toggle = useCallback(() => {
    setValue((currentValue) => !currentValue)
  }, [])

  return [value, toggle]
}
