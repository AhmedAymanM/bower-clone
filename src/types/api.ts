// README: for convince and efficacy I used record type for any param is not
// used in the task
export type Project = Record<string, unknown> & {
  name: string
  description: string
  repository_url: string
  stars: number
  latest_stable_release_number: string
}
