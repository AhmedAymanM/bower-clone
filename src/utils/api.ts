import { env } from '@/utils/env'

export async function request<TResponse>(
  url: string,
  init?: RequestInit
): Promise<{ data: TResponse; totalCount?: number }> {
  const response = await fetch(env.API_URL + url, init)

  if (!response.ok) throw new Error(response.statusText)

  let totalCount
  if (response.headers.has('Total')) {
    totalCount = +response.headers.get('Total')!
  }

  const data = await response.json()
  return { data, totalCount }
}

export const api = {
  get: <TResponse>(url: string) => request<TResponse>(url),
}
