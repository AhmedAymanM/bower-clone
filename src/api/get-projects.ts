import { type Project } from '@/types/api'
import { api } from '@/utils/api'
import { env } from '@/utils/env'
import { generateSearchQuery } from '@/utils/generateSearchQuery'
import { useAsync } from '@/hooks/useAsync'
import { useCallback } from 'react'

type sortEnum =
  | 'rank'
  | 'stars'
  | 'dependents_count'
  | 'dependent_repos_count'
  | 'latest_release_published_at'
  | 'contributions_count'
  | 'created_at'

type GetProjectsProps = {
  sort?: sortEnum
  q?: string
  page?: number
}

const RESULTS_PER_PAGE = 5

export const getProjects = ({ sort, q, page }: GetProjectsProps = {}) => {
  const searchQuery = generateSearchQuery({
    api_key: env.API_KEY,
    per_page: RESULTS_PER_PAGE.toString(),
    page: page?.toString(),
    sort,
    q,
  })

  return api.get<Project[]>('/search' + searchQuery)
}

export const useProjects = ({ sort, q, page }: GetProjectsProps = {}) => {
  const memoizedGetProjects = useCallback(
    () => getProjects({ sort, q, page }),
    [page, q, sort]
  )

  return useAsync<Project[]>(memoizedGetProjects)
}
