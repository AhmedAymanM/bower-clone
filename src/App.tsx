import { useEffect, type FC } from 'react'

import { Header } from '@/components/features/header'
import { ContentLayout } from '@/components/layouts/content-layout'
import { Sidebar } from '@/components/features/sidebar'
import { SearchInput } from '@/components/ui/search-input'
import { getProjects } from '@/api/get-projects'

export const App: FC = () => {
  useEffect(() => {
    getProjects().then((data) => console.log({ data }))
  }, [])

  return (
    <>
      <Header />

      <ContentLayout className="my-4">
        <Sidebar />
        <section className="flex-auto">
          <SearchInput
            placeholder="Search..."
            autoComplete="off"
            type="search"
            // eslint-disable-next-line jsx-a11y/no-autofocus
            autoFocus
          />
        </section>
      </ContentLayout>
    </>
  )
}
