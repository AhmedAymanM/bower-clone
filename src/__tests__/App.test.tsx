// README: Ideally I would avoid this separation if not needed and write
// integration/e2e tests with cypress or similar with help of stubbed/mocked
// data using tools like msw (depending on our API)

import { render } from '@testing-library/react'
import { fireEvent, screen } from '@testing-library/dom'

import { STATUS } from '@/hooks/useAsync'
import { type Project } from '@/types/api'

import { App, type AppProps } from '../App'

const stubProjects: Project[] = [
  {
    description:
      'TypeScript is a language for application scale JavaScript development',
    homepage: 'https://www.typescriptlang.org/',
    latest_stable_release_number: '5.5.3',
    name: 'typescript',
    repository_url: 'https://github.com/microsoft/TypeScript',
    stars: 99231,
  },
  {
    description: 'TypeScript definitions for node',
    latest_stable_release_number: '20.14.10',
    name: '@types/node',
    repository_url: 'https://github.com/DefinitelyTyped/DefinitelyTyped',
    stars: 47771,
  },
  {
    description: 'React package for working with the DOM.',
    latest_stable_release_number: '18.3.1',
    name: 'react-dom',
    repository_url: 'https://github.com/facebook/react-dom',
    stars: 225291,
  },
  {
    description: 'React is a JavaScript library for building user interfaces.',
    latest_stable_release_number: '18.3.1',
    name: 'react',
    repository_url: 'https://github.com/facebook/react',
    stars: 225291,
  },
  {
    description: 'Promise based HTTP client for the browser and node.js',
    latest_stable_release_number: '1.7.2',
    name: 'axios',
    repository_url: 'https://github.com/axios/axios',
    stars: 104653,
  },
]

const generateCommonProps = (props?: Partial<AppProps>) => ({
  searchQuery: '',
  sort: undefined,
  handleSortChange: vi.fn(),
  handleSearchChange: vi.fn(),
  setCurrentPage: vi.fn(),
  projects: {
    data: {
      data: stubProjects,
      totalCount: 8541079,
    },
    status: STATUS.SUCCESS,
    error: undefined,
  },
  ...props,
})

describe('App component', () => {
  test('renders the App component correctly', () => {
    const commonProps = generateCommonProps()
    const { asFragment } = render(<App {...commonProps} />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('SearchInput is rendered and triggered on search input interaction', () => {
    const commonProps = generateCommonProps()
    const { getByPlaceholderText } = render(<App {...commonProps} />)

    const searchInput = getByPlaceholderText(/search.../i)
    expect(searchInput).toBeInTheDocument()

    fireEvent.change(searchInput, { target: { value: 'test-search' } })
    expect(commonProps.handleSearchChange).toHaveBeenCalled()
    expect(searchInput).toHaveValue('test-search')
  })
  test('onSortChange is triggered on sort options interaction', () => {
    const commonProps = generateCommonProps()
    const { getByTestId } = render(<App {...commonProps} />)

    const sortOptions = getByTestId('sort-projects')
    expect(sortOptions).toBeInTheDocument()

    fireEvent.change(sortOptions, { target: { value: 'rank' } })
    expect(commonProps.handleSortChange).toHaveBeenCalled()
  })
  test('onPageChange is triggered on pagination interaction', () => {
    const commonProps = generateCommonProps()

    const { getByLabelText } = render(<App {...commonProps} />)

    const nextPageButton = getByLabelText(/next page/i)
    fireEvent.click(nextPageButton)
    expect(commonProps.setCurrentPage).toHaveBeenCalled()
  })
  test('renders the ProjectsCards grid with 5 cards', () => {
    const commonProps = generateCommonProps()
    render(<App {...commonProps} />)
    const projectsCardsGrid = screen.getByTestId(/cards-grid/i)
    expect(projectsCardsGrid).toBeInTheDocument()
    expect(projectsCardsGrid.childElementCount).toEqual(5)
  })
  test('renders project cards placeholder if projects are loading', () => {
    const commonProps = generateCommonProps({
      projects: {
        data: undefined,
        status: STATUS.LOADING,
        error: undefined,
      },
    })
    const { getAllByTestId } = render(<App {...commonProps} />)

    const projectCardPlaceholder = getAllByTestId('project-card-placeholder')
    expect(projectCardPlaceholder.length).toEqual(5)
  })
  test('renders error message if projects are errored', () => {
    const commonProps = generateCommonProps({
      projects: {
        data: undefined,
        status: STATUS.ERROR,
        error: new Error('Something went wrong'),
      },
    })
    const { getByText } = render(<App {...commonProps} />)

    const errorMessage = getByText(/Something went wrong/i)
    expect(errorMessage).toBeInTheDocument()
  }),
    test('renders messages if no projects are found', () => {
      const commonProps = generateCommonProps({
        projects: {
          data: undefined,
          status: STATUS.SUCCESS,
          error: undefined,
        },
      })
      const { getByText } = render(<App {...commonProps} />)

      const noProjectsMessage = getByText(/no projects were found/i)
      expect(noProjectsMessage).toBeInTheDocument()
    })
})
