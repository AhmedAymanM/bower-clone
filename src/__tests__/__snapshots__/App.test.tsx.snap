// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders the App component correctly 1`] = `
<DocumentFragment>
  <header
    class="bg-header-bg px-2 py-4 text-header-text md:py-8"
  >
    <div
      class="mx-auto max-w-5xl px-2 md:px-4 flex items-center gap-2 md:gap-10"
    >
      <a
        class="text-blue hover:underline transition-transform duration-200 hover:scale-110"
        href="bower.io"
        rel="noopener"
        target="_blank"
      >
        <img
          alt="Bower Logo"
          class="w-10 md:w-[140px]"
          src="/src/assets/BowerLogo.svg"
        />
      </a>
      <div
        class="font-bold tracking-tight"
      >
        <h1
          class="text-3xl md:text-6xl"
        >
          Bower Search
        </h1>
        <h4
          class="hidden text-xl md:block"
        >
          Powered by 
          <a
            class="text-blue hover:underline"
            href="https://libraries.io/"
            rel="noopener"
            target="_blank"
          >
            libraries.io
          </a>
        </h4>
      </div>
    </div>
  </header>
  <div
    class="mx-auto max-w-5xl px-2 md:px-4 flex flex-col gap-4 md:flex-row my-4"
  >
    <section
      class="flex flex-col gap-4"
    >
      <button
        class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray m-auto w-full max-w-64 md:hidden"
      >
        Menu
      </button>
      <nav
        class="flex flex-col font-bold text-lg md:w-[200px]"
      >
        <ul
          class="hidden flex-col gap-2 text-center md:flex md:text-left"
        >
          <li>
            <a
              class="hover:underline text-blue-light"
              href="https://bower.io/"
              rel="noopener"
              target="_blank"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="hover:underline text-blue-light"
              href="https://bower.io/docs/creating-packages"
              rel="noopener"
              target="_blank"
            >
              Creating Packages
            </a>
          </li>
          <li>
            <a
              class="hover:underline text-blue-light"
              href="https://bower.io/docs/api"
              rel="noopener"
              target="_blank"
            >
              API
            </a>
          </li>
          <li>
            <a
              class="hover:underline text-blue-light"
              href="https://bower.io/docs/config"
              rel="noopener"
              target="_blank"
            >
              configuration
            </a>
          </li>
          <li>
            <a
              class="hover:underline text-blue-light"
              href="https://bower.io/docs/pluggable-resolvers"
              rel="noopener"
              target="_blank"
            >
              Pluggable Resolvers
            </a>
          </li>
          <li>
            <a
              class="hover:underline text-blue-light"
              href="https://bower.io/docs/bower-tools"
              rel="noopener"
              target="_blank"
            >
              Tools
            </a>
          </li>
          <li>
            <a
              class="hover:underline text-blue-light"
              href="https://bower.io/docs/about"
              rel="noopener"
              target="_blank"
            >
              About
            </a>
          </li>
        </ul>
      </nav>
    </section>
    <section
      class="flex flex-auto flex-col gap-4"
    >
      <div
        class="flex flex-col items-center gap-2 sm:flex-row"
      >
        <input
          autocomplete="off"
          class="w-full rounded border border-gray p-2.5 px-2 text-lg font-medium leading-6 outline-none placeholder:italic focus:border-blue-light"
          placeholder="Search..."
          type="search"
          value=""
        />
        <select
          class="w-full rounded border border-gray p-2.5 px-2 leading-6 text-black-accent outline-none focus:border-blue-light basis-1/3"
          data-testid="sort-projects"
          name="Sort by"
        >
          <option
            value=""
          >
            Sort by
          </option>
          <option
            value="stars"
          >
            Stars
          </option>
          <option
            value="rank"
          >
            Rank
          </option>
          <option
            value="created_at"
          >
            Creation date
          </option>
          <option
            value="latest_release_published_at"
          >
            Latest release
          </option>
          <option
            value="contributions_count"
          >
            Contributions
          </option>
          <option
            value="dependents_count"
          >
            Dependents
          </option>
          <option
            value="dependent_repos_count"
          >
            Dependent repos
          </option>
        </select>
      </div>
      <section
        class="grid grid-cols-1 content-start justify-center justify-items-center gap-x-3 gap-y-4 lg:grid-cols-2"
        data-testid="cards-grid"
      >
        <a
          class="flex max-h-fit min-h-28 w-full flex-col overflow-hidden rounded-md border border-gray bg-white hover:border-blue-light"
          data-testid="project-card"
          href="https://github.com/microsoft/TypeScript"
          rel="noopener"
          target="_blank"
        >
          <div
            class="flex flex-wrap items-center justify-between bg-white-accent p-2"
          >
            <div
              class="flex items-baseline gap-1 truncate"
            >
              <h2
                class="pb-0.5 text-lg font-bold"
              >
                typescript
              </h2>
              <p
                class="text-xs text-black/60"
              >
                by microsoft
              </p>
            </div>
            <div
              class="flex items-center gap-1.5"
            >
              <p
                class="text-xs text-black/60"
              >
                <span
                  aria-label="number of stars"
                  role="img"
                >
                  ⭐
                </span>
                 99k
              </p>
              <p
                class="text-xs text-black/60 rounded bg-black/10 px-1 py-px"
              >
                v5.5.3
              </p>
            </div>
          </div>
          <div
            class="my-auto p-2"
          >
            <p
              class="line-clamp-2 text-black-accent"
            >
              TypeScript is a language for application scale JavaScript development
            </p>
          </div>
        </a>
        <a
          class="flex max-h-fit min-h-28 w-full flex-col overflow-hidden rounded-md border border-gray bg-white hover:border-blue-light"
          data-testid="project-card"
          href="https://github.com/DefinitelyTyped/DefinitelyTyped"
          rel="noopener"
          target="_blank"
        >
          <div
            class="flex flex-wrap items-center justify-between bg-white-accent p-2"
          >
            <div
              class="flex items-baseline gap-1 truncate"
            >
              <h2
                class="pb-0.5 text-lg font-bold"
              >
                @types/node
              </h2>
              <p
                class="text-xs text-black/60"
              >
                by DefinitelyTyped
              </p>
            </div>
            <div
              class="flex items-center gap-1.5"
            >
              <p
                class="text-xs text-black/60"
              >
                <span
                  aria-label="number of stars"
                  role="img"
                >
                  ⭐
                </span>
                 48k
              </p>
              <p
                class="text-xs text-black/60 rounded bg-black/10 px-1 py-px"
              >
                v20.14.10
              </p>
            </div>
          </div>
          <div
            class="my-auto p-2"
          >
            <p
              class="line-clamp-2 text-black-accent"
            >
              TypeScript definitions for node
            </p>
          </div>
        </a>
        <a
          class="flex max-h-fit min-h-28 w-full flex-col overflow-hidden rounded-md border border-gray bg-white hover:border-blue-light"
          data-testid="project-card"
          href="https://github.com/facebook/react-dom"
          rel="noopener"
          target="_blank"
        >
          <div
            class="flex flex-wrap items-center justify-between bg-white-accent p-2"
          >
            <div
              class="flex items-baseline gap-1 truncate"
            >
              <h2
                class="pb-0.5 text-lg font-bold"
              >
                react-dom
              </h2>
              <p
                class="text-xs text-black/60"
              >
                by facebook
              </p>
            </div>
            <div
              class="flex items-center gap-1.5"
            >
              <p
                class="text-xs text-black/60"
              >
                <span
                  aria-label="number of stars"
                  role="img"
                >
                  ⭐
                </span>
                 225k
              </p>
              <p
                class="text-xs text-black/60 rounded bg-black/10 px-1 py-px"
              >
                v18.3.1
              </p>
            </div>
          </div>
          <div
            class="my-auto p-2"
          >
            <p
              class="line-clamp-2 text-black-accent"
            >
              React package for working with the DOM.
            </p>
          </div>
        </a>
        <a
          class="flex max-h-fit min-h-28 w-full flex-col overflow-hidden rounded-md border border-gray bg-white hover:border-blue-light"
          data-testid="project-card"
          href="https://github.com/facebook/react"
          rel="noopener"
          target="_blank"
        >
          <div
            class="flex flex-wrap items-center justify-between bg-white-accent p-2"
          >
            <div
              class="flex items-baseline gap-1 truncate"
            >
              <h2
                class="pb-0.5 text-lg font-bold"
              >
                react
              </h2>
              <p
                class="text-xs text-black/60"
              >
                by facebook
              </p>
            </div>
            <div
              class="flex items-center gap-1.5"
            >
              <p
                class="text-xs text-black/60"
              >
                <span
                  aria-label="number of stars"
                  role="img"
                >
                  ⭐
                </span>
                 225k
              </p>
              <p
                class="text-xs text-black/60 rounded bg-black/10 px-1 py-px"
              >
                v18.3.1
              </p>
            </div>
          </div>
          <div
            class="my-auto p-2"
          >
            <p
              class="line-clamp-2 text-black-accent"
            >
              React is a JavaScript library for building user interfaces.
            </p>
          </div>
        </a>
        <a
          class="flex max-h-fit min-h-28 w-full flex-col overflow-hidden rounded-md border border-gray bg-white hover:border-blue-light"
          data-testid="project-card"
          href="https://github.com/axios/axios"
          rel="noopener"
          target="_blank"
        >
          <div
            class="flex flex-wrap items-center justify-between bg-white-accent p-2"
          >
            <div
              class="flex items-baseline gap-1 truncate"
            >
              <h2
                class="pb-0.5 text-lg font-bold"
              >
                axios
              </h2>
              <p
                class="text-xs text-black/60"
              >
                by axios
              </p>
            </div>
            <div
              class="flex items-center gap-1.5"
            >
              <p
                class="text-xs text-black/60"
              >
                <span
                  aria-label="number of stars"
                  role="img"
                >
                  ⭐
                </span>
                 105k
              </p>
              <p
                class="text-xs text-black/60 rounded bg-black/10 px-1 py-px"
              >
                v1.7.2
              </p>
            </div>
          </div>
          <div
            class="my-auto p-2"
          >
            <p
              class="line-clamp-2 text-black-accent"
            >
              Promise based HTTP client for the browser and node.js
            </p>
          </div>
        </a>
      </section>
      <div
        class="flex items-center gap-2 self-center"
      >
        <button
          aria-label="Previous page"
          class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
          disabled=""
        >
          ⬅︎
        </button>
        <span
          class="font-bold"
        >
          1 / 1,708,216
        </span>
        <button
          aria-label="Next page"
          class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
        >
          ➡︎
        </button>
      </div>
    </section>
  </div>
  <footer
    class="mt-12 px-2 text-sm"
  >
    <div
      class="mx-auto max-w-5xl px-2 md:px-4 border-t border-gray py-3 pl-0"
    >
      <a
        class="text-blue hover:underline my-4"
        href="https://github.com/bower/bower.github.io/issues"
        rel="noopener"
        target="_blank"
      >
        Help improve these docs. Open an issue or pull request.
      </a>
    </div>
  </footer>
</DocumentFragment>
`;
